pkg/virt-handler/vm.go

```diff
@@ -240,28 +239,78 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 		// This is needed to be backwards compatible with vmi's which have status interfaces
 		// with the name not being set
-		if len(vmi.Status.Interfaces) == 1 && vmi.Status.Interfaces[0].Name == "" {
+		if len(domain.Spec.Devices.Interfaces) == 0 && len(vmi.Status.Interfaces) == 1 && vmi.Status.Interfaces[0].Name == "" {
 			for _, network := range vmi.Spec.Networks {
 				if network.NetworkSource.Pod != nil {
 					vmi.Status.Interfaces[0].Name = network.Name
 				}
 			}
 		}
 
-		interfacesByName := make(map[string]int)
-		for i, existingInterface := range vmi.Status.Interfaces {
-			interfacesByName[existingInterface.Name] = i
-		}
+		if len(domain.Spec.Devices.Interfaces) > 0 || len(domain.Status.Interfaces) > 0 {
+			// This calculates the vmi.Status.Interfaces based on the following data sets:
+			// - vmi.Status.Interfaces - previously calculated interfaces, this can contains data
+			//   set in the controller (pod IP) which can not be deleted, unless overridden by Qemu agent
+			// - domain.Spec - interfaces form the Spec
+			// - domain.Status.Interfaces - interfaces reported by guest agent (emtpy if Qemu agent not running)
+			newInterfaces := []v1.VirtualMachineInstanceNetworkInterface{}
+
+			existingInterfaceStatusByName := map[string]v1.VirtualMachineInstanceNetworkInterface{}
+			for _, existingInterfaceStatus := range vmi.Status.Interfaces {
+				if existingInterfaceStatus.Name != "" {
+					existingInterfaceStatusByName[existingInterfaceStatus.Name] = existingInterfaceStatus
+				}
+			}
 
-		for _, domainInterface := range domain.Spec.Devices.Interfaces {
-			if i, exists := interfacesByName[domainInterface.Alias.Name]; exists {
-				vmi.Status.Interfaces[i].MAC = domainInterface.MAC.MAC
-			} else {
-				vmi.Status.Interfaces = append(vmi.Status.Interfaces, v1.VirtualMachineInstanceNetworkInterface{MAC: domainInterface.MAC.MAC, Name: domainInterface.Alias.Name})
+			domainInterfaceStatusByMac := map[string]api.InterfaceStatus{}
+			for _, domainInterfaceStatus := range domain.Status.Interfaces {
+				domainInterfaceStatusByMac[domainInterfaceStatus.Mac] = domainInterfaceStatus
+			}
+
+			// Iterate through all domain.Spec interfaces
+			for _, domainInterface := range domain.Spec.Devices.Interfaces {
+				interfaceMAC := domainInterface.MAC.MAC
+				var newInterface v1.VirtualMachineInstanceNetworkInterface
+
+				if existingInterface, exists := existingInterfaceStatusByName[domainInterface.Alias.Name]; exists && domainInterface.Alias.Name != "" {
+					// Reuse previously calculated interface from vmi.Status.Interfaces, updating the MAC from domain.Spec
+					// Only interfaces defined in domain.Spec are handled here
+					newInterface = existingInterface
+					newInterface.MAC = interfaceMAC
+				} else {
+					// If not present in vmi.Status.Interfaces, create a new one based on domain.Spec
+					newInterface = v1.VirtualMachineInstanceNetworkInterface{
+						MAC:  interfaceMAC,
+						Name: domainInterface.Alias.Name,
```

Comment:
So in case domainInterface.Alias.Name == "", you still report it? How that happens?